packages/gatsby/src/bootstrap/load-themes/index.ts

```diff
@@ -0,0 +1,249 @@
+import { createRequireFromPath } from "gatsby-core-utils"
+import path from "path"
+import { mergeGatsbyConfig } from "../../utils/merge-gatsby-config"
+import Bluebird from "bluebird"
+import _ from "lodash"
+import { preferDefault } from "../prefer-default"
+import { getConfigFile } from "../get-config-file"
+import { resolvePlugin } from "../load-plugins/load"
+import reporter from "gatsby-cli/lib/reporter"
+import Debug from "debug"
+import { GatsbyConfig, IPluginRefOptions, PluginRef } from "../../../"
+
+const debug = Debug(`gatsby:load-themes`)
+
+interface IResolvedTheme {
+  themeName: string
+  themeConfig: GatsbyConfig
+  themeSpec: PluginRef
+  themeDir: string
+  parentDir: string
+  configFilePath: string
+}
+// get the gatsby-config file for a theme
+const resolveTheme = async (
+  themeSpec: PluginRef,
+  configFileThatDeclaredTheme: string,
+  isMainConfig = false,
+  rootDir: string
+): Promise<IResolvedTheme> => {
+  const themeName =
+    typeof themeSpec === `string` ? themeSpec : themeSpec.resolve
+  let themeDir = ``
+  try {
+    const scopedRequire = createRequireFromPath(
+      `${rootDir}/:internal:`
+    ) as NodeRequire
+    // theme is an node-resolvable module
+    themeDir = path.dirname(scopedRequire.resolve(themeName))
+  } catch (e) {
+    let pathToLocalTheme = ``
+
+    // only try to look for local theme in main site
+    // local themes nested in other themes is potential source of problems:
+    // because those are not hosted by npm, there is potential for multiple
+    // local themes with same name that do different things and name being
+    // main identifier that Gatsby uses right now, it's safer not to support it for now.
+    if (isMainConfig) {
+      pathToLocalTheme = path.join(path.resolve(`.`), `plugins`, themeName)
+      // is a local plugin OR it doesn't exist
+      try {
+        const { resolve } = resolvePlugin(themeName, rootDir)
+        themeDir = resolve
+      } catch (localErr) {
+        // catch shouldn't be empty :shrug:
+      }
+    }
+
+    if (!themeDir) {
+      const nodeResolutionPaths = module.paths.map((p: string): string =>
+        path.join(p, themeName)
+      )
```

Comment:
Hi @ascorbic, I actually feel like typing functions such as these ones are counter-productive. I know the guidelines state that "Every function should type its input and output (even if it’s inferrable)". May I get some insights as to why Gatsby opted for this kind of rule?